tests_func/load/mongo_load.go

```diff
@@ -0,0 +1,247 @@
+	Ctx      string                       `json:"ctx"`
+	MakeLogs bool                         `json:"makelogs"`
+	Cmds     []map[string]json.RawMessage `json:"cmds"`
+	callback func(client *mongo.Client) (interface{}, error)
+}
+
+func parseTransactionOp(cd map[string]json.RawMessage) (*MongoRunTransactionOp, error) {
+	var res MongoRunTransactionOp
+	bmrco, err := json.Marshal(cd)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse opts for commad: %+v", err)
+	}
+	err = json.Unmarshal(bmrco, &res)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse opts for commad: %+v", err)
+	}
+	var fns []func(client *mongo.Client)
+	for _, cmd := range res.Cmds {
+		commandDoc, err := parseCommandOp(cmd)
+		if err != nil {
+			return nil, err
+		}
+		fns = append(fns, func(client *mongo.Client) {
+			mongoRunCommandOp(client, commandDoc)
+		})
+	}
+	res.callback = func(client *mongo.Client) (interface{}, error) {
+		for _, f := range fns {
+			f(client)
+		}
+		return nil, nil
+	}
+	return &res, nil
+}
+
+func (mw *MongoWorker) mongoRunTransaction(client *mongo.Client, op *MongoRunTransactionOp) {
+	mw.fs = append(mw.fs, func() {
+		session, err := client.StartSession()
+		if op.MakeLogs {
+			if err != nil {
+				tracelog.InfoLogger.Printf("Cannot start session: %+v", err)
+			} else {
+				tracelog.InfoLogger.Printf("Successful execution of transaction with %+v argument", op.Cmds)
+			}
+		}
+		if err != nil {
+			<-mw.ch
+			mw.wg.Done()
+			return
+		}
+		defer session.EndSession(context.Background())
+		_, err = session.WithTransaction(contextByName[op.Ctx], func(sessCtx mongo.SessionContext) (interface{}, error) {
+			return op.callback(client)
+		})
+		if op.MakeLogs {
+			if err != nil {
+				tracelog.InfoLogger.Printf("Failed execution of transaction: %+v", err)
+			} else {
+				tracelog.InfoLogger.Printf("Successful execution of transaction with %+v argument", op.Cmds)
+			}
+		}
+		<-mw.ch
+		mw.wg.Done()
+	})
+}
+
+func (mw *MongoWorker) Run() {
+	for _, f := range mw.fs {
+		mw.ch <- struct{}{}
+		mw.wg.Add(1)
+		go f()
+	}
+	mw.wg.Wait()
+}
+
+func multiplyInArray(value string, cnt int) []string {
+	var res []string
+	for cnt > 0 {
+		res = append(res, value)
+		cnt--
+	}
+	return res
+}
+
+func main() {
+
+	err := generatePatronsFromFile("./tests_func/load/mongo_load_config.json")
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	cli, err := mongo.NewClient(options.Client().ApplyURI("mongodb://localhost:27017"))
+	if err != nil {
+		fmt.Printf("cannot create client to mongodb: %v", err)
+		return
+	}
+	err = cli.Connect(context.TODO())
+	if err != nil {
+		fmt.Printf("cannot connect to mongodb: %v", err)
+	}
+	err = cli.Ping(context.TODO(), nil)
+	if err != nil {
+		fmt.Printf("cannot ping mongodb basse: %v", err)
+		return
+	}
+
+	dat, err := ioutil.ReadFile("./tests_func/load/mongo_load_patron_1.json")
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	mw := CreateMongoWorker(3)
+	err = mw.AddMongoOp(cli, string(dat))
```

Comment:
Our test documents can have a very large size (up to 16mb each), so we can not hold all "bullets" in memory.
I see an option: 

1. read bullets one by one from JSON file using stream interface (https://golang.org/pkg/encoding/json/#Decoder.Decode)
2. apply them one by one to worker's pool via buffered channel

Reply:
Fixed